src/analyzer/protocol/http/HTTP.cc

```diff
@@ -1331,6 +1322,60 @@ void HTTP_Analyzer::ReplyMade(bool interrupted, const char* msg) {
         reply_state = EXPECT_REPLY_LINE;
 }
 
+void HTTP_Analyzer::HTTP_Upgrade() {
+    // Upgraded connection that switches immediately - e.g. websocket.
+
+    // Lookup an analyzer tag in the HTTP::upgrade_analyzer table.
+    static const auto& upgrade_analyzers = id::find_val<TableVal>("HTTP::upgrade_analyzers");
+
+    auto upgrade_protocol_val = make_intrusive<StringVal>(upgrade_protocol);
+    auto v = upgrade_analyzers->Find(upgrade_protocol_val);
+    if ( ! v ) {
+        // If not found, try the all lower version, too.
+        auto lower_upgrade_protocol = util::strtolower(upgrade_protocol);
+        upgrade_protocol_val = make_intrusive<StringVal>(lower_upgrade_protocol);
+        v = upgrade_analyzers->Find(upgrade_protocol_val);
+    }
+
+    if ( v ) {
+        auto analyzer_tag_val = cast_intrusive<EnumVal>(v);
+        DBG_LOG(DBG_ANALYZER, "Found %s in HTTP::upgrade_analyzers for %s",
+                analyzer_tag_val->GetType<EnumType>()->Lookup(analyzer_tag_val->AsEnum()),
+                upgrade_protocol_val->CheckString());
+        auto analyzer_tag = analyzer_mgr->GetComponentTag(analyzer_tag_val.get());
+        auto* analyzer = analyzer_mgr->InstantiateAnalyzer(analyzer_tag, Conn());
+        if ( analyzer ) {
+            AddChildAnalyzer(analyzer);
+
+            // Hmm, for the WebSocket analyzer we may want to drain EventMgr
+            // here so that analyzer configuration completes before continuing.
+            //
+            // Practically this only matters if there's protocol data available
+            // directly following the protocol upgrade. And anyway, this data
+            // would be buffered in the ContentLine analyzer that we're removing
+            // just below.
```

Comment:
As we already touched in conversation I'm a bit worried about this because I can see WebSocket-centric tooling that considers all of the upgrade boilerplate that just needs to be shoved down the pipes to get going on the actual frame exchange. Or put differently, if it's an immediately clear evasion we should see what we can do to counter it.

Reply:
> Or put differently, if it's an immediately clear evasion we should see what we can do to counter it.

Well, it is. Not sure we'd see it in the wild, but can add the "weird" and attempt to drain the ContentLine analyzer.

Reply:
I've implemented the logic to handle this properly for the res ponder side and added a weird for when it happens. Also extended the CONNECT handling to do the same.

For the originator side it's a different story. If one feeds data completely independently into the analyzer, we would need to add request/response synchronization as we wouldn't know whether to parse the next bit of data as a pipelined HTTP request, or the ugraded/connect protocol.